validation handler benchmark tests

shravan-achar opened the pull request:
**What this PR does / why we need it**:

- The PR performs a benchmark of Gatekeeper's validating handler. 

- A sample of PSP constraint templates are pre-loaded into OPA. Each subtest pre-loads M constraints where M varies from 5 to 100000. For each of the M constraints, number of admission requests that can be processed in 1 second is measured.

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
- These tests do not measure the HTTP roundtrip to/from kube-api-server

Current results:
cores: 8
mem: 16GB

> goos: darwin
> goarch: amd64
> pkg: github.com/open-policy-agent/gatekeeper/pkg/webhook
> BenchmarkValidationHandler/5-8 	     129	   9385359 ns/op
> 
> BenchmarkValidationHandler/10-8         	      78	  18034209 ns/op
> 
> BenchmarkValidationHandler/50-8         	      15	  78782760 ns/op
> 
> BenchmarkValidationHandler/100-8        	       7	 208199975 ns/op
> 
> BenchmarkValidationHandler/200-8        	       4	 329748445 ns/op
> 
> BenchmarkValidationHandler/1000-8       	       1	1846845599 ns/op
> 
> BenchmarkValidationHandler/2000-8       	       1	3729658819 ns/op
> 
> BenchmarkValidationHandler/10000-8      	       1	17484500470 ns/op
> 
> BenchmarkValidationHandler/20000-8      	       1	35564250359 ns/op
> 
> BenchmarkValidationHandler/100000-8     	       1	165198614593 ns/op

With memory stats

> goos: darwin
> goarch: amd64
> pkg: github.com/open-policy-agent/gatekeeper/pkg/webhook
> 
> BenchmarkValidationHandler/5-8 	     141	   7852219 ns/op	 3658307 B/op	   56034 allocs/op
> 
> BenchmarkValidationHandler/10-8         	      79	  15501142 ns/op	 7221695 B/op	  110534 allocs/op
> 
> BenchmarkValidationHandler/50-8         	      14	  73609200 ns/op	35739040 B/op	  547383 allocs/op
> 
> BenchmarkValidationHandler/100-8        	       7	 148565541 ns/op	71472198 B/op	 1095124 allocs/op
> 
> BenchmarkValidationHandler/200-8        	       4	 305889273 ns/op	143454862 B/op	 2200257 allocs/op
> 
> BenchmarkValidationHandler/1000-8       	       1	1401957112 ns/op	689022800 B/op	10466498 allocs/op
> 
> BenchmarkValidationHandler/2000-8       	       1	2874750865 ns/op	1377815096 B/op	20930929 allocs/op
> 
> BenchmarkValidationHandler/10000-8      	       1	14572899917 ns/op	6888856152 B/op	104646480 allocs/op
> 
> BenchmarkValidationHandler/20000-8      	       1	29800708818 ns/op	13777700688 B/op	209290439 allocs/op
> 
> BenchmarkValidationHandler/100000-8     	       1	161775454897 ns/op	69075147392 B/op	1046442550 allocs/op

shravan-achar commented:
I signed it

shravan-achar commented:
@maxsmythe @ritazh @shomron  I have added some tests to measure number of admission requests that can be handled by Gatekeeper when varying number of constraints are pre-loaded into OPA. Please take a look to see if this can be helpful. I have also pasted one set of results in the PR description. 

The output follows the following format:
{{Test Name}}/{{Number of Constraints}}-{{CPU cores}} {{Number of iterations}} {{Average time for each iteration in ns/op}}

For ex: 
`BenchmarkValidationHandler/5-8 129 9385359 ns/op` would mean that a maximum of 129 pod admission requests were reviewed by GK's validation handler in 1 second. Average time for each iteration was 9.3 microseconds with 5 constraints were pre-loaded. 

The tests run atleast 1 iteration even if takes more than 1 second to complete. This is seen for `BenchmarkValidationHandler/1000-8 1 1846845599 ns/op` where 1000 constraints were pre-loaded and a single iteration took 1.8 seconds. 

Please let me know if you have any questions.

maxsmythe commented on pkg/webhook/policy_benchmark_test.go:
```diff
@@ -0,0 +1,301 @@
+}
+func buildTemplatePath(path string) string {
+	var b strings.Builder
+	b.WriteString(path)
```
`path.Join()`?

shravan-achar replied:
This will be fixed

maxsmythe commented on pkg/webhook/policy_benchmark_test.go:
```diff
@@ -0,0 +1,301 @@
+	return string(out)
+}
+
+// libraryPathList contains a random set of library paths with resources
```
This is not currently random, should the docstring be updated? Would we want it random, or would we want consistent test inputs to promote comparability?

shravan-achar replied:
Docstring should be updated. I was going with consistent test inputs to promote comparability. It would be hard to characterize the results with random inputs

shravan-achar commented:
@maxsmythe - Thanks for taking a look. I am happy to add more tests in another PR.  
Could you clarify what testing is being referred to here
> When we are doing kind testing in the template itself

maxsmythe commented:
re: taking a look, absolutely, happy to look :)

WRT "When we are doing kind testing in the template itself" I am referring to when there is an explicit kind check in the `violation` rule as opposed to using match semantics, e.g.:

```
// only deny Pods named bad
violation[{"msg": "not allowed"}] {
   input.review.kind.kind == "Pod"
   input.review.object.metadata.name == "bad"
}
```

shravan-achar commented:
@maxsmythe - I have changed the test based on earlier feedback. Separated out templates, constraints and resources into separate test folders. Modified helper method signatures.

One can curate a list of templates, constraints and objects to define a new test. Non-applicable constraints or violating/non-violating requests could be added as separate test cases. Please give it a look when you get a chance. 
cc: @ritazh @shomron

maxsmythe commented on pkg/webhook/policy_benchmark_test.go:
```diff
@@ -0,0 +1,332 @@
+		// number of constraints to load
+		load []int
+	}{
+		"psp": {
```
nit: we should name this based off the description... maybe `psp: 100% violations` or similar

shravan-achar replied:
done

shravan-achar replied:
changed

maxsmythe commented on pkg/webhook/policy_benchmark_test.go:
```diff
@@ -0,0 +1,332 @@
+			rand.Seed(time.Now().UnixNano())
+
+			// create T templates
+			err = addTemplates(opa, ctList)
```
Would it be hard to add a benchmark for template/constraint loading here? Looks like we could just call separate b.Run()s?

shravan-achar replied:
It might be better to add template and constraint loading as a separate benchmark test. It would enable us to use different data set for each , and would be easier to read and extend.
Let me know your thoughts on this.

I can add another benchmark test for these two operations (loading of constraints and templates), which is not too hard.

maxsmythe replied:
If you think it's best done as a different test and doing so doesn't majorly change the design of what's already here, we can add it later.

shravan-achar replied:
The overall process will be very similar. We would need to read M templates or generate them, and the operation under test `opa.AddTemplate` would be put under a b.Run() block. Go benchmark will automatically run it for multiple iterations and generate a report of the average time per operation (along with memory usage if so desired). M can be varied from 1 to 1000 (any number desired). Same for constraints.

This resource was helpful to me: https://golang.org/pkg/testing/#hdr-Benchmarks 

Adding another test should not impact the design of this one.

maxsmythe commented on pkg/webhook/testdata/psp-constraints/host-filesystem-constraint.yaml:
```diff
@@ -0,0 +1,13 @@
+apiVersion: constraints.gatekeeper.sh/v1beta1
```
we probably want to put each test case in its own directory. If there are common files, we can split them out into a shared directory somewhere.

shravan-achar replied:
+1 on having a separate directory for each test case. Change is on the way.

shravan-achar replied:
Changed

shravan-achar commented on pkg/webhook/policy_benchmark_test.go:
```diff
@@ -0,0 +1,330 @@
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/meta"
```
done

shravan-achar commented:
Thank you for the review on this. Very much liked writing this part. Love the Gatekeeper project. I would be interested in continuing to contribute to this project if that's alright. If there is an interest in other benchmarking tests that came up during this review, I would be happy to write them. Even otherwise, I would be happy to add bug fixes/improvements.

maxsmythe commented:
Thank you for the contribution! I'm glad it was fun :)

Of course we'd love it if you continued to contribute! The more the merrier!

I was definitely interested in the potential for benchmarking the adding of constraint templates and the caching of data. Templates is a known pain point that we should address some day.

Other than that, let us know if there are any issues you're interested in working on or designs you'd like to address and we'll do our best to get you set up.
